apps/speed.c

```diff
@@ -2867,38 +2971,145 @@ int speed_main(int argc, char **argv)
+                            && mode_op != EVP_CIPH_GCM_SIV_MODE) {
+                            if (!EVP_CIPHER_CTX_ctrl(loopargs[k].ctx,
+                                                     EVP_CTRL_AEAD_SET_TAG,
+                                                     TAG_LEN, NULL)) {
+                                BIO_printf(bio_err,
+                                           "\nFailed to set tag length\n");
+                                dofail();
+                                exit(1);
+                            }
+                        }
+                        if (!EVP_CipherInit_ex(loopargs[k].ctx, NULL, NULL,
+                                               loopargs[k].key, aead_iv, -1)) {
+                            BIO_printf(bio_err, "\nFailed to set the key\n");
+                            dofail();
+                            exit(1);
+                        }
+                        /* Set total length of input. Only required for CCM */
+                        if (mode_op == EVP_CIPH_CCM_MODE) {
+                            if (!EVP_EncryptUpdate(loopargs[k].ctx, NULL,
+                                                   &outlen, NULL,
+                                                   lengths[testnum])) {
+                                BIO_printf(bio_err,
+                                           "\nCouldn't set input text length\n");
+                                dofail();
+                                exit(1);
+                            }
+                        }
+                        if (aead) {
+                            if (!EVP_EncryptUpdate(loopargs[k].ctx, NULL,
+                                                   &outlen, aad, sizeof(aad))) {
+                                BIO_printf(bio_err,
+                                           "\nCouldn't insert AAD when encrypting\n");
+                                dofail();
+                                exit(1);
+                            }
+                        }
+                        if (!EVP_EncryptUpdate(loopargs[k].ctx, loopargs[k].buf,
+                                               &outlen, loopargs[k].buf,
+                                               lengths[testnum])) {
+                            BIO_printf(bio_err,
+                                       "\nFailed to to encrypt the data\n");
+                            dofail();
+                            exit(1);
+                        }
+
+                        if (!EVP_EncryptFinal_ex(loopargs[k].ctx,
+                                                 loopargs[k].buf, &outlen)) {
+                            BIO_printf(bio_err,
+                                       "\nFailed finalize the encryption\n");
+                            dofail();
+                            exit(1);
+                        }
+
+                        if (!EVP_CIPHER_CTX_ctrl(loopargs[k].ctx, EVP_CTRL_AEAD_GET_TAG,
```

Comment:
Same comment here as well.

Reading long functions can be quite challenging. It's best to keep functions as small as possible, adhering to the single responsibility principle. This not only aids in quick understanding but also makes maintenance easier.

Reply:
@shahsb Thanks a million for the feedback. I agree with all proposed changes. In fact, I was working on refactoring the code and adding support for chacha as well, but the intention was to create another PR as this one has been approved. What do you think?

Reply:
Now that this PR is ready to merge, please don't make further changes to it. Any further changes are best addressed in a follow-on PR. If you want to refactor, maybe do that as one PR, before fixing ChaCha in a subsequent PR (smaller, more focussed PRs are easier and quicker to review - a query on one aspect doesn't hold up something that is only tangentially related)

Reply:
Yes, I agree with addressing this as part of separate PRs. 

Also, I 100% agree on the fact that smaller & focused PRs are easier & quick to review. @tom-cosgrove-arm  - Thanks, Good point!

Reply:
> @shahsb Thanks a million for the feedback. I agree with all proposed changes. In fact, I was working on refactoring the code and adding support for chacha as well, but the intention was to create another PR as this one has been approved. What do you think?

Yes, Thanks @zelda923 - I really appreciate your work, Good job!